pkg/controller/shoot/shoot_quota_control.go

```diff
@@ -0,0 +1,195 @@
+	"github.com/gardener/gardener/pkg/logger"
+	"github.com/gardener/gardener/pkg/operation/common"
+	"github.com/gardener/gardener/pkg/utils"
+	"k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+func (c *Controller) shootQuotaAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return
+	}
+	c.shootQuotaQueue.Add(key)
+}
+
+func (c *Controller) shootQuotaDelete(obj interface{}) {
+	shoot, ok := obj.(*gardenv1beta1.Shoot)
+	if shoot == nil || !ok {
+		return
+	}
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return
+	}
+	c.shootQuotaQueue.Done(key)
+}
+
+func (c *Controller) reconcileShootQuotaKey(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	shoot, err := c.shootLister.Shoots(namespace).Get(name)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if err := c.quotaControl.CheckQuota(shoot, key); err != nil {
+		c.shootQuotaQueue.AddAfter(key, 2*time.Minute)
+		return nil
+	}
+	c.shootQuotaQueue.AddAfter(key, time.Hour)
+	return nil
+}
+
+// QuotaControlInterface implements the control logic for quota management of Shoots. It is implemented as an interface to allow
+// for extensions that provide different semantics. Currently, there is only one implementation.
+type QuotaControlInterface interface {
+	CheckQuota(shoot *gardenv1beta1.Shoot, key string) error
+}
+
+// NewDefaultQuotaControl returns a new instance of the default implementation of QuotaControlInterface
+// which implements the semantics for controling the quota handling of Shoot resources.
+func NewDefaultQuotaControl(k8sGardenClient kubernetes.Client, k8sGardenInformers gardeninformers.Interface) QuotaControlInterface {
+	return &defaultQuotaControl{k8sGardenClient, k8sGardenInformers}
+}
+
+type defaultQuotaControl struct {
+	k8sGardenClient    kubernetes.Client
+	k8sGardenInformers gardeninformers.Interface
+}
+
+func (c *defaultQuotaControl) CheckQuota(shootObj *gardenv1beta1.Shoot, key string) error {
+	var (
+		clusterLifeTime     *int
+		quotaReferences     []v1.ObjectReference
+		shootExpirationTime time.Time
+
+		operationID = utils.GenerateRandomString(8)
+		shoot       = shootObj.DeepCopy()
+		shootLogger = logger.NewShootLogger(logger.Logger, shoot.Name, shoot.Namespace, operationID)
+	)
+
+	switch shoot.Spec.Cloud.SecretBindingRef.Kind {
+	case "PrivateSecretBinding":
+		psb, err := c.k8sGardenInformers.
+			PrivateSecretBindings().
+			Lister().
+			PrivateSecretBindings(shoot.Namespace).
+			Get(shoot.Spec.Cloud.SecretBindingRef.Name)
+		if err != nil {
+			return err
+		}
+		quotaReferences = psb.Quotas
+	case "CrossSecretBinding":
+		csb, err := c.k8sGardenInformers.
+			CrossSecretBindings().
+			Lister().
+			CrossSecretBindings(shoot.Namespace).
```

Comment:
This should be `shoot.Spec.Cloud.SecretBindingRef.Namespace`.

Reply:
I thought the both kind of binding have to be in the same namespace as the shoot.
@mliepold mentioned that here in the story https://github.com/gardener/gardener/issues/81

The shoot resource dependency admission controller is also checking for bindings in the shoot namespace. https://github.com/gardener/gardener/blob/master/plugin/pkg/global/resourcereferencemanager/admission.go

Reply:
Agree with @dkistner. In the current design with Private- and CrossSecrectBindings, there is no namespace in the secretBindingRef.
Only in the proposal from @mvladev that would remove the secret bindings there is a namespace in the secretBindingRef.